### PR TITLE
Add version command to driver

### PR DESCRIPTION
**Pull request by turneram**
This PR adds a command (version) and an option (--version or -v) to the driver. Both print the current working version of MIGraphX.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +52,12 @@ add_library(migraphx
     normalize_attributes.cpp
     normalize_ops.cpp
 )
+file(WRITE include/migraphx/version.h
```
Comment (pfultz2) [inline src/CMakeLists.txt:55]: This shouldn't write to the source tree. Instead, there should be a version.h.in and this file should be configured by cmake.

File: src/driver/main.cpp [hunk 2]
```diff
@@ -423,6 +424,16 @@ struct verify : command<verify>
     }
 };
 
+struct version : command<version>
+{
+    compiler c;
```
Comment (pfultz2) [inline src/driver/main.cpp:429]: This is never used and can be removed.

File: src/CMakeLists.txt (hunk 1 above)
Comment (turneram) [reply src/CMakeLists.txt]: Fixed

File: src/driver/main.cpp (hunk 2 above)
Comment (turneram) [reply src/driver/main.cpp]: Removed

**Review by kahmed10 — APPROVED** [review]

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -167,6 +168,7 @@ target_link_libraries(migraphx PRIVATE -lstdc++fs)
 endif()
 
 target_include_directories(migraphx SYSTEM PUBLIC $<BUILD_INTERFACE:${HALF_INCLUDE_DIR}>)
+include_directories(${CMAKE_CURRENT_BINARY_DIR})
```
Comment (pfultz2) [inline src/CMakeLists.txt:171]: The include should be added to the target: `target_include_directories(migraphx PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)`.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -52,6 +52,7 @@ add_library(migraphx
     normalize_attributes.cpp
     normalize_ops.cpp
 )
+configure_file(version.h.in version.h)
```
Comment (pfultz2) [inline src/CMakeLists.txt:55]: This should be under the migraphx directory: `configure_file(version.h.in include/migraphx/version.h)`.

Comment (pfultz2) [reply src/CMakeLists.txt:55]: Also the header should be installed, which can be done by updating the `rocm_install_targets` below and adding the new include directory:

```cmake
rocm_install_targets(
  TARGETS migraphx
  INCLUDE
    ${CMAKE_CURRENT_SOURCE_DIR}/include
    ${CMAKE_CURRENT_BINARY_DIR}/include
)
```

File: src/CMakeLists.txt (hunk 3 above)
Comment (turneram) [reply src/CMakeLists.txt]: Fixed

File: src/CMakeLists.txt (hunk 4 above)
Comment (turneram) [reply src/CMakeLists.txt]: Fixed

File: src/driver/main.cpp [hunk 5]
```diff
@@ -423,6 +424,15 @@ struct verify : command<verify>
     }
 };
 
+struct version : command<version>
+{
+    void parse(argument_parser) {}
```
Comment (pfultz2) [inline src/driver/main.cpp:429]: This should probably be `void parse(const argument_parser&) {}` to avoid the tidy warning.

**Comment by scxiao** [comment]
@turneram @pfultz2 I am wondering whether we should always print the version info in the first line no matter what we execute..

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -167,6 +169,7 @@ target_link_libraries(migraphx PRIVATE -lstdc++fs)
 endif()
 
 target_include_directories(migraphx SYSTEM PUBLIC $<BUILD_INTERFACE:${HALF_INCLUDE_DIR}>)
+target_include_directories(migraphx PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
```
Comment (pfultz2) [inline src/CMakeLists.txt:172]: Actually, I think this line is not needed as `rocm_install_targets` will add the include directory paths.

**Review by pfultz2 — APPROVED** [review]

**Comment by mvermeulen** [comment]
@turneram assume you didn't change msgpack.cpp, why are the format tests failing now?
Cc: @pfultz2

**Review by scxiao — APPROVED** [review]

**Comment by pfultz2** [comment]
The .o files should be removed from this PR.

**Comment by pfultz2** [comment]
> assume you didn't change msgpack.cpp, why are the format tests failing now?

I dont see a formatting failure in msgpack. These are format diffs shown:

```
432,433c432
 sh -c clang-format-5.0 -style=file ./src/driver/main.cpp | diff - ./src/driver/main.cpp 
sh -c clang-format-5.0 -style=file ./src/driver/models.hpp | diff - ./src/driver/models.hpp 
<         std::cout << "MIGraphX Version: " << MIGRAPHX_VERSION_MAJOR << "." << MIGRAPHX_VERSION_MINOR
<                   << std::endl;
---
>         std::cout << "MIGraphX Version: " << MIGRAPHX_VERSION_MAJOR << "." << MIGRAPHX_VERSION_MINOR << std::endl;
518,519c517,518
<         std::string version_str = "MIGraphX Version: " + std::to_string(MIGRAPHX_VERSION_MAJOR) +
<                                   "." + std::to_string(MIGRAPHX_VERSION_MINOR);
---
>         std::string version_str = "MIGraphX Version: " + 
>             std::to_string(MIGRAPHX_VERSION_MAJOR) + "." + std::to_string(MIGRAPHX_VERSION_MINOR);
521,524c520,521
<         ap(nullptr,
<            {"-v", "--version"},
<            ap.help("Show MIGraphX version"),
<            ap.show_help(version_str));
---
>         ap(nullptr, {"-v", "--version"}, ap.help("Show MIGraphX version"), ap.show_help(version_str));
> 
sh -c clang-format-5.0 -style=file ./src/version.h.in | diff - ./src/version.h.in 
1,2c1,2
< #define MIGRAPHX_VERSION_MAJOR @PROJECT_VERSION_MAJOR @
< #define MIGRAPHX_VERSION_MINOR @PROJECT_VERSION_MINOR @
---
> #define MIGRAPHX_VERSION_MAJOR @PROJECT_VERSION_MAJOR@
> #define MIGRAPHX_VERSION_MINOR @PROJECT_VERSION_MINOR@
```

So main.cpp should be formatted, but we should skip formatting version.h.in(as adding the space wont configure correctly in cmake) using `// clang-format off`.

**Pull request merged**